stdlib/option.mli

```diff
@@ -0,0 +1,64 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Option values.
+
+    @since NEXT_RELEASE *)
+
+(** {1:options Options} *)
+
+type 'a t = 'a option
+(** The type for option values. Either [Some v] or [None]. *)
+
+val some : 'a -> 'a option
+(** [some v] is [Some v]. *)
+
+val evict : none:'a -> 'a option -> 'a
+(** [evict ~none o] is [none] if [o] is [None] and [v] if [o] is [Some v]. *)
```

Comment:
On one hand, this name is typical "bunzli", which makes it very amusing. On the other hand, it's completely unhelpful. `may` is sometimes use for this, but I don't find it particularly clear either. I quite like Container's convention which uses the `_or` prefix for operation that define a default value. You then `get_or` and `map_or`.

Reply:
The signature in cute should also be considered:

```
value : 'a option -> default:'a -> 'a
```

In my opinion, both core and containers chose better names here.

Reply:
I did use `Option.value` in the past but eventually dismissed: I don't find it compelling since `option`s are also values themselves. 

Regarding the label the advantage of using `~none` rather than `~default` is that you can consistently use it everywhere to denote what happens in this case. E.g. `Option.to_result`, `Format.pp_print_option`, or whatever may be added in the future that needs a special argument for `None`.

Reply:
`extract` perhaps? `evict` reads very weirdly to a native English speaker, I would assume it would destroy something rather than extracting the wrapped value.

Reply:
`evict` doesn't make much sense to me either

Reply:
What is so strange about `Option.evict` ? It means you are [evicting](https://www.merriam-webster.com/dictionary/evict) or expelling the option; but then I'm not a native english speaker. 

As for myself I don't find `Option.extract` to be meaningful you are not extracting the option you are extracting a value out of the option.

In any case please provide full signatures with doc strings for alternate suggestions. For now the only other good candidate to my eyes would be @drup's suggestion albeit with a `none` label rather than `default` for the reasons mentioned above, i.e. :
```ocaml
val get_or ~none:'a -> 'a t -> 'a
(** [get_or ~none o] is [none] if [o] is [None] and [v] if [o] is [Some v].
```

Reply:
Here are a few alternatives names all with the signature given above to be read in an example context. I might post an @mseri style poll at some point: 

```ocaml
let v = Option.evict ~none:x (Map.find_opt k m)
let v = Option.get_or ~none:x (Map.find_opt k m)
let v = Option.remove ~none:x (Map.find_opt k m) 
let v = Option.expel ~none:x (Map.find_opt k m)
let v = Option.elim ~none:x (Map.find_opt k m)
```

I'm quite tempted by simply pushing `Option.remove` though.

Reply:
My vote would be for 1. `Option.unwrap` (currently not in the list) 2. `Option.get_or`
The keyword argument though should be `default` rather than `none` IMO, as the former is much clearer.

Reply:
The Haskell equivalent is called [`fromMaybe`](https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Maybe.html#v:fromMaybe), which suggests  `from_option` or `of_option`.

Reply:
@yallop doesn't `Option.{of,from}_option` read very weirdly to you ? It seems at odds with the OCaml `M.of` conventions.

@bluddy I explained why the argument is `none` (note that this pattern also works for e.g. `result`, see the various labellings in the API of [`rresult`](http://erratique.ch/software/rresult/doc/Rresult.R.html)) do you suggest: 

1. To be inconsistent and use `~default` only in that function and `~none` in the others ?
2. To use `~default` in all of these other functions; I doubt that it is a great idea.

If `none` is mentioned in the name though we get could then label with `use`, something like:

```ocaml
let v = Option.on_none ~use:x (Map.find_opt k m) 
```

Reply:
@dbuenzli: it reads okay to me, although I can see how it seems at odds with the other `of` functions.

One more possibility (in case there aren't enough under consideration already):  the SML basis uses [`getOPT`](http://sml-family.org/Basis/option.html#OPTION:SIG:SPEC); we could use `Option.get` here, and drop the function that raises `Invalid_argument`.

Reply:
Having a raising and a non-raising version would be good, yes.

As for this:

> What is so strange about `Option.evict`?

To a native speaker, eviction has completely incorrect connotations. One evicts a person who is late on their rent, for example, and the connotation is not of extracting something to examine it but rather of banishment. The word really isn't correct. One is not evicting the value (which sounds like what you would do when removing an entry from a hash table), one is fetching the value. Words like fetch, get, peek, look, examine, extract, etc. come to mind. Evict really means "banish", and you are not banishing the value, you're getting it.